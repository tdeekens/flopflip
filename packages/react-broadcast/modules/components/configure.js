import PropTypes from 'prop-types';
import React from 'react';
import { FlagsSubscription } from '@flopflip/react';
import Broadcast from './broadcast';

export const FLAGS_CHANNEL = '@flopflip';

export default class Configure extends React.Component {
  static displayName = 'ConfigureFlopflip';

  static propTypes = {
    children: PropTypes.node,
    clientSideId: PropTypes.string.isRequired,
    shouldInitialize: PropTypes.bool,
    shouldChangeUserContext: PropTypes.bool,
    defaultFlags: PropTypes.object,
    user: PropTypes.shape({
      key: PropTypes.string,
    }),
  };

  static defaultProps = {
    children: null,
    user: {},
    defaultFlags: {},
    shouldInitialize: true,
    shouldChangeUserContext: false,
  };

  state = {
    flags: {},
    status: { isReady: false },
  };

  handleUpdateFlags = flags => {
    this.setState(prevState => ({
      flags: {
        ...prevState.flags,
        ...flags,
      },
    }));
  };

  handleUpdateStatus = status => {
    this.setState({ status });
  };

  render() {
    return (
      <FlagsSubscription
        clientSideId={this.props.clientSideId}
        user={this.props.user}
        defaultFlags={this.props.defaultFlags}
        shouldInitialize={this.props.shouldInitialize}
        shouldChangeUserContext={this.props.shouldChangeUserContext}
        onUpdateStatus={this.handleUpdateStatus}
        onUpdateFlags={this.handleUpdateFlags}
      >
        <Broadcast channel={FLAGS_CHANNEL} value={this.state.flags}>
          {this.props.children ? (
            React.Children.only(this.props.children)
          ) : null}
        </Broadcast>
      </FlagsSubscription>
    );
  }
}
