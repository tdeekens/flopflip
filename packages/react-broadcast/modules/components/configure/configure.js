// @flow

import type {
  Flags,
  Adapter,
  AdapterArgs,
  AdapterState,
  AdapterStatus,
} from '@flopflip/types';

import React, { PureComponent, type ComponentType, type Node } from 'react';
import memoize from 'lodash.memoize';
import createReactContext, { type Context } from 'create-react-context';
import { ConfigureAdapter } from '@flopflip/react';

type Props = {
  children: Node,
  shouldDeferAdapterConfiguration?: boolean,
  defaultFlags?: Flags,
  adapterArgs: AdapterArgs,
  adapter: Adapter,
};
type State = {
  flags: Flags,
};

export const FlagsContext: Context<Flags> = createReactContext({});

const createAdapterArgs = memoize(
  (adapterArgs, handleUpdateStatus, handleUpdateFlags) => ({
    ...adapterArgs,
    onStatusStateChange: handleUpdateStatus,
    onFlagsStateChange: handleUpdateFlags,
  })
);

export default class Configure extends PureComponent<Props, State> {
  static displayName = 'ConfigureFlopflip';

  static defaultProps = {
    children: null,
    defaultFlags: {},
    shouldDeferAdapterConfiguration: false,
  };

  state: { flags: Flags } = {
    flags: {},
  };

  handleUpdateFlags = (flags: Flags): void => {
    this.setState(prevState => ({
      flags: {
        ...prevState.flags,
        ...flags,
      },
    }));
  };

  handleUpdateStatus = (status: AdapterStatus): void =>
    this.setState(prevState => ({ ...prevState, ...status }));

  render(): Node {
    return (
      <ConfigureAdapter
        adapter={this.props.adapter}
        adapterArgs={createAdapterArgs(
          this.props.adapterArgs,
          this.handleUpdateStatus,
          this.handleUpdateFlags
        )}
        defaultFlags={this.props.defaultFlags}
        shouldDeferAdapterConfiguration={
          this.props.shouldDeferAdapterConfiguration
        }
      >
        <FlagsContext.Provider value={this.state.flags}>
          {this.props.children
            ? React.Children.only(this.props.children)
            : null}
        </FlagsContext.Provider>
      </ConfigureAdapter>
    );
  }
}
