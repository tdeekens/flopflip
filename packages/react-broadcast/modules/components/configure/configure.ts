import { Flags, Adapter, AdapterArgs, AdapterStatus } from '@flopflip/types';

import React from 'react';
import PropTypes from 'prop-types';
import { ConfigureAdapter } from '@flopflip/react';
import { FlagsContext } from '../flags-context';

type Props = {
  children?: Node;
  shouldDeferAdapterConfiguration?: boolean;
  defaultFlags?: Flags;
  adapterArgs: AdapterArgs;
  adapter: Adapter;
};
type State = {
  flags: Flags;
};

/**
 * NOTE:
 *    This function can not be memoized otherwise it will
 *    rewrire different `ConfigureFlopflip`s to the same change
 *    handlers as these functions seem to be referentially equal even
 *    across different instances of React components due to hoisting.
 */
const createAdapterArgs = (
  adapterArgs: AdapterArgs,
  handleUpdateStatus: (status: AdapterStatus) => void,
  handleUpdateFlags: (flags: Flags) => void
): AdapterArgs => ({
  ...adapterArgs,
  onStatusStateChange: handleUpdateStatus,
  onFlagsStateChange: handleUpdateFlags,
});

export default class Configure extends React.PureComponent<Props, State> {
  static displayName = 'ConfigureFlopflip';

  static defaultProps = {
    children: null,
    defaultFlags: {},
    shouldDeferAdapterConfiguration: false,
  };

  static propTypes = {
    children: PropTypes.element,
    defaultFlags: PropTypes.object,
    shouldDeferAdapterConfiguration: PropTypes.bool,
  };

  state: { flags: Flags } = {
    flags: {},
  };

  isUnmounted: boolean = false;

  componentDidMount() {
    this.isUnmounted = false;
  }

  componentWillUnmount() {
    this.isUnmounted = true;
  }

  handleUpdateFlags = (flags: Flags): void => {
    !this.isUnmounted &&
      this.setState(prevState => ({
        flags: {
          ...prevState.flags,
          ...flags,
        },
      }));
  };

  handleUpdateStatus = (status: AdapterStatus): void => {
    !this.isUnmounted &&
      this.setState(prevState => ({ ...prevState, ...status }));
  };

  render(): React.Node {
    return (
      <ConfigureAdapter
        adapter={this.props.adapter}
        adapterArgs={createAdapterArgs(
          this.props.adapterArgs,
          this.handleUpdateStatus,
          this.handleUpdateFlags
        )}
        defaultFlags={this.props.defaultFlags}
        shouldDeferAdapterConfiguration={
          this.props.shouldDeferAdapterConfiguration
        }
      >
        <FlagsContext.Provider value={this.state.flags}>
          {this.props.children
            ? React.Children.only(this.props.children)
            : null}
        </FlagsContext.Provider>
      </ConfigureAdapter>
    );
  }
}
