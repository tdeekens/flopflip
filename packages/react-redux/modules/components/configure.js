import PropTypes from 'prop-types';
import React from 'react';
import { connect } from 'react-redux';
import { FlagsSubscription } from '@flopflip/react';
import { updateStatus, updateFlags } from './../ducks';

export class Configure extends React.Component {
  static displayName = 'ConfigureFlopflip';
  static propTypes = {
    children: PropTypes.node,
    clientSideId: PropTypes.string.isRequired,
    user: PropTypes.shape({
      key: PropTypes.string,
    }),
    defaultFlags: PropTypes.object,
    shouldInitialize: PropTypes.bool,

    // Connected
    handleUpdateStatus: PropTypes.func.isRequired,
    handleUpdateFlags: PropTypes.func.isRequired,
  };

  static defaultProps = {
    children: null,
    user: {},
    defaultFlags: {},
    shouldInitialize: true,
  };

  render() {
    return (
      <FlagsSubscription
        clientSideId={this.props.clientSideId}
        user={this.props.user}
        defaultFlags={this.props.defaultFlags}
        shouldInitialize={this.props.shouldInitialize}
        onUpdateStatus={this.props.handleUpdateStatus}
        onUpdateFlags={this.props.handleUpdateFlags}
      >
        {this.props.children ? React.Children.only(this.props.children) : null}
      </FlagsSubscription>
    );
  }
}

const mapDispatchToProps = {
  handleUpdateStatus: updateStatus,
  handleUpdateFlags: updateFlags,
};

export default connect(null, mapDispatchToProps)(Configure);
