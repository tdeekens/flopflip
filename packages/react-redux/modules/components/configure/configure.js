// @flow

import type {
  Flags,
  AdapterArgs,
  AdapterState,
  AdapterStatus,
} from '@flopflip/types';

import React, { PureComponent, type ComponentType, type Node } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { ConfigureAdapter } from '@flopflip/react';
import { updateStatus, updateFlags } from './../../ducks';

type Props = {
  children?: Node,
  shouldDeferAdapterConfiguration?: boolean,
  defaultFlags?: Flags,
  adapterArgs: AdapterArgs,
  adapter: mixed,
};
type ConnectedProps = {
  handleUpdateStatus: () => void,
  handleUpdateFlags: () => void,
};
type State = {
  flags: Flags,
};

/**
 * NOTE:
 *    This function can not be memoized otherwise it will
 *    rewrire different `ConfigureFlopflip`s to the same change
 *    handlers as these functions seem to be referentially equal even
 *    across different instances of React components due to hoisting.
 */
const createAdapterArgs = (
  adapterArgs,
  handleUpdateStatus,
  handleUpdateFlags
) => ({
  ...adapterArgs,
  onStatusStateChange: handleUpdateStatus,
  onFlagsStateChange: handleUpdateFlags,
});

export class Configure extends PureComponent<Props & ConnectedProps, State> {
  static displayName = 'ConfigureFlopflip';

  static defaultProps = {
    children: null,
    defaultFlags: {},
    shouldDeferAdapterConfiguration: false,
  };

  static propTypes = {
    children: PropTypes.element,
    defaultFlags: PropTypes.object,
    shouldDeferAdapterConfiguration: PropTypes.bool,
  };

  render(): Node {
    return (
      <ConfigureAdapter
        adapter={this.props.adapter}
        adapterArgs={createAdapterArgs(
          this.props.adapterArgs,
          this.props.handleUpdateStatus,
          this.props.handleUpdateFlags
        )}
        defaultFlags={this.props.defaultFlags}
        shouldDeferAdapterConfiguration={
          this.props.shouldDeferAdapterConfiguration
        }
      >
        {this.props.children ? React.Children.only(this.props.children) : null}
      </ConfigureAdapter>
    );
  }
}

const mapDispatchToProps: ConnectedProps = {
  handleUpdateStatus: updateStatus,
  handleUpdateFlags: updateFlags,
};

/* istanbul ignore next */
export default connect(
  null,
  mapDispatchToProps
)(Configure);
