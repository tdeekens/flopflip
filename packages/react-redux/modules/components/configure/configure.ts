// @flow

import {
  Flags,
  Adapter,
  AdapterArgs,
  AdapterStatus,
} from '@flopflip/types';
import { UpdateFlagsAction, UpdateStatusAction } from '../../types'

import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { ConfigureAdapter } from '@flopflip/react';
import { updateStatus, updateFlags } from './../../ducks';

type Props = {
  children?: React.Node,
  shouldDeferAdapterConfiguration?: boolean,
  defaultFlags?: Flags,
  adapterArgs: AdapterArgs,
  adapter: Adapter,
};
type ConnectedProps = {
  handleUpdateStatus: (status: AdapterStatus) => UpdateStatusAction,
  handleUpdateFlags: (flags: Flags) => UpdateFlagsAction,
};
type State = {
  flags: Flags,
};

/**
 * NOTE:
 *    This function can not be memoized otherwise it will
 *    rewrire different `ConfigureFlopflip`s to the same change
 *    handlers as these functions seem to be referentially equal even
 *    across different instances of React components due to hoisting.
 */
const createAdapterArgs = (
  adapterArgs: AdapterArgs,
  handleUpdateStatus: (status: AdapterStatus) => UpdateStatusAction,
  handleUpdateFlags: (flags: Flags) => UpdateFlagsAction
): AdapterArgs => ({
  ...adapterArgs,
  onStatusStateChange: handleUpdateStatus,
  onFlagsStateChange: handleUpdateFlags,
});

export class Configure extends React.PureComponent<Props & ConnectedProps, State> {
  static displayName = 'ConfigureFlopflip';

  static defaultProps = {
    children: null,
    defaultFlags: {},
    shouldDeferAdapterConfiguration: false,
  };

  static propTypes = {
    children: PropTypes.element,
    defaultFlags: PropTypes.object,
    shouldDeferAdapterConfiguration: PropTypes.bool,
  };

  render(): React.Node {
    return (
      <ConfigureAdapter
        adapter={this.props.adapter}
        adapterArgs={createAdapterArgs(
          this.props.adapterArgs,
          this.props.handleUpdateStatus,
          this.props.handleUpdateFlags
        )}
        defaultFlags={this.props.defaultFlags}
        shouldDeferAdapterConfiguration={
          this.props.shouldDeferAdapterConfiguration
        }
      >
        {this.props.children ? React.Children.only(this.props.children) : null}
      </ConfigureAdapter>
    );
  }
}

const mapDispatchToProps: ConnectedProps = {
  handleUpdateStatus: updateStatus,
  handleUpdateFlags: updateFlags,
};

/* istanbul ignore next */
export default connect(
  null,
  mapDispatchToProps
)(Configure);
