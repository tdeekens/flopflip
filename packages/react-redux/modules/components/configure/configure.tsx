import React from 'react';
import { connect } from 'react-redux';
import { ConfigureAdapter } from '@flopflip/react';
import {
  Flags,
  Adapter,
  AdapterArgs,
  AdapterStatus,
  ConfigureAdapterChildren,
} from '@flopflip/types';
import { UpdateFlagsAction, UpdateStatusAction } from '../../types';
import { updateStatus, updateFlags } from './../../ducks';

type Props = {
  children?: ConfigureAdapterChildren;
  shouldDeferAdapterConfiguration?: boolean;
  defaultFlags?: Flags;
  adapterArgs: AdapterArgs;
  adapter: Adapter;
};
type ConnectedProps = {
  handleUpdateStatus: (status: AdapterStatus) => UpdateStatusAction;
  handleUpdateFlags: (flags: Flags) => UpdateFlagsAction;
};
type State = {
  flags: Flags;
};

/**
 * NOTE:
 *    This function can not be memoized otherwise it will
 *    rewrire different `ConfigureFlopflip`s to the same change
 *    handlers as these functions seem to be referentially equal even
 *    across different instances of React components due to hoisting.
 */
const createAdapterArgs = (
  adapterArgs: AdapterArgs,
  handleUpdateStatus: (status: AdapterStatus) => UpdateStatusAction,
  handleUpdateFlags: (flags: Flags) => UpdateFlagsAction
): AdapterArgs => ({
  ...adapterArgs,
  onStatusStateChange: handleUpdateStatus,
  onFlagsStateChange: handleUpdateFlags,
});

export class Configure extends React.PureComponent<
  Props & ConnectedProps,
  State
> {
  static displayName = 'ConfigureFlopflip';

  static defaultProps = {
    defaultFlags: {},
    shouldDeferAdapterConfiguration: false,
  };

  render(): React.ReactNode {
    return (
      <ConfigureAdapter
        adapter={this.props.adapter}
        adapterArgs={createAdapterArgs(
          this.props.adapterArgs,
          this.props.handleUpdateStatus,
          this.props.handleUpdateFlags
        )}
        defaultFlags={this.props.defaultFlags}
        shouldDeferAdapterConfiguration={
          this.props.shouldDeferAdapterConfiguration
        }
      >
        {this.props.children}
      </ConfigureAdapter>
    );
  }
}

const mapDispatchToProps: ConnectedProps = {
  handleUpdateStatus: updateStatus,
  handleUpdateFlags: updateFlags,
};

/* istanbul ignore next */
export default connect(
  null,
  mapDispatchToProps
)(Configure);
