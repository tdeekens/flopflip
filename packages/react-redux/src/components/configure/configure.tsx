import { ConfigureAdapter, useAdapterSubscription } from '@flopflip/react';
import type {
  TAdapter,
  TConfigureAdapterChildren,
  TConfigureAdapterProps,
  TFlags,
} from '@flopflip/types';
import React from 'react';

import { useUpdateFlags, useUpdateStatus } from '../../hooks';

type BaseProps = {
  children?: TConfigureAdapterChildren;
  shouldDeferAdapterConfiguration?: boolean;
  defaultFlags?: TFlags;
};
type Props<AdapterInstance extends TAdapter> = BaseProps &
  TConfigureAdapterProps<AdapterInstance>;

const defaultProps: Pick<
  BaseProps,
  'defaultFlags' | 'shouldDeferAdapterConfiguration'
> = {
  defaultFlags: {},
  shouldDeferAdapterConfiguration: false,
};

const Configure = <AdapterInstance extends TAdapter>(
  props: Props<AdapterInstance>
) => {
  const adapterIdentifiers = [props.adapter.id];
  const handleUpdateFlags = useUpdateFlags({adapterIdentifiers});
  const handleUpdateStatus = useUpdateStatus();

  useAdapterSubscription(props.adapter);

  return (
    <ConfigureAdapter
      adapter={props.adapter}
      adapterArgs={props.adapterArgs}
      defaultFlags={props.defaultFlags}
      shouldDeferAdapterConfiguration={props.shouldDeferAdapterConfiguration}
      onFlagsStateChange={handleUpdateFlags}
      onStatusStateChange={handleUpdateStatus}
    >
      {props.children}
    </ConfigureAdapter>
  );
};

Configure.displayName = 'ConfigureFlopflip';
Configure.defaultProps = defaultProps;

export default Configure;
