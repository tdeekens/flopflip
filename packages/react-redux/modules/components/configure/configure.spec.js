import React from 'react';
import { shallow } from 'enzyme';
import { ConfigureAdapter } from '@flopflip/react';
import { Configure } from './configure';

const ChildComponent = () => <div />;
const createTestProps = custom => ({
  adapter: {
    configure: jest.fn(),
    reconfigure: jest.fn(),
  },
  adapterArgs: {
    fooId: 'foo-id',
  },

  // HoC
  handleUpdateFlags: jest.fn(),
  handleUpdateStatus: jest.fn(),

  ...custom,
});

describe('rendering', () => {
  let props;
  let wrapper;

  beforeEach(() => {
    props = createTestProps();
    wrapper = shallow(<Configure {...props} />);
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a `<ConfigureAdapter>`', () => {
    expect(wrapper).toRender(ConfigureAdapter);
  });

  describe('with `children`', () => {
    let props;

    beforeEach(() => {
      props = createTestProps();

      wrapper = shallow(
        <Configure {...props}>
          <ChildComponent />
        </Configure>
      );
    });

    it('should render `children`', () => {
      expect(wrapper).toRender(ChildComponent);
    });
  });

  describe('`of <ConfigureAdapter />`', () => {
    let configureAdapterWrapper;

    beforeEach(() => {
      configureAdapterWrapper = wrapper.find(ConfigureAdapter);
    });

    it('should receive `adapterArgs`', () => {
      expect(configureAdapterWrapper).toHaveProp(
        'adapterArgs',
        expect.objectContaining({})
      );
    });

    it('should receive `onStatusStateChange` and `onFlagsStateChange` in `adapterArgs`', () => {
      expect(configureAdapterWrapper).toHaveProp(
        'adapterArgs',
        expect.objectContaining({
          onStatusStateChange: props.handleUpdateStatus,
          onFlagsStateChange: props.handleUpdateFlags,
        })
      );
    });

    it('should receive `defaultFlags`', () => {
      expect(configureAdapterWrapper).toHaveProp(
        'defaultFlags',
        wrapper.prop('defaultFlags')
      );
    });
  });
});

describe('statics', () => {
  describe('displayName', () => {
    it('should be set to `ConfigureFlopflip`', () => {
      expect(Configure.displayName).toEqual('ConfigureFlopflip');
    });
  });

  describe('defaultProps', () => {
    it('should default `defaultFlags` to an empty object', () => {
      expect(Configure.defaultProps.defaultFlags).toEqual({});
    });

    it('should default `children` to `null`', () => {
      expect(Configure.defaultProps.children).toBe(null);
    });

    it('should default `shouldDeferAdapterConfiguration` to `true`', () => {
      expect(Configure.defaultProps.shouldDeferAdapterConfiguration).toBe(
        false
      );
    });
  });
});
