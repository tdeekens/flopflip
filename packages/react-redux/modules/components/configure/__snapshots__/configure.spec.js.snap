// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering should match snapshot 1`] = `
<ConfigureAdapter
  adapter={
    Object {
      "configure": [MockFunction],
      "reconfigure": [MockFunction],
    }
  }
  adapterArgs={
    Object {
      "fooId": "foo-id",
      "onFlagsStateChange": [MockFunction],
      "onStatusStateChange": [MockFunction],
    }
  }
  defaultFlags={Object {}}
  render={null}
  shouldDeferAdapterConfiguration={false}
/>
`;
